### PR TITLE
OS spec

### PR DESCRIPTION
**Pull request by jpwilliams**
## Description

OS SDK specification.

### CODE REVIEW

File: docs/SDK_SPEC.md [hunk 1]
```diff
@@ -0,0 +1,1148 @@
...
+   * A unique ID used to idempotently process a given event payload.
+   *
+   * A developer can set this when sending events to ensure that the event is
+   * only processed once; if an event with the same ID is sent again, it will
+   * not trigger any functions.
+   */
+	id: string;
+
+  /**
+   * A unique IDfor the type of event. We recommend using lowercase dot
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:88]:
```suggestion
   * A unique ID for the type of event. We recommend using lowercase dot
```

File: docs/SDK_SPEC.md [hunk 2]
```diff
@@ -0,0 +1,1148 @@
...
+3. If the SDK is intending to use Inngest Cloud [4.2] and no `X-Inngest-Signature` header has been given, the SDK MUST reject the request and return a `500 Internal Server Error` response.
+4. The SDK SHOULD then extract the `t` timestamp from the `X-Inngest-Signature` header and ensure it is a time within the last 5 minutes according to the SDK’s known time.
+
+If this is checked and the timestamp is not within this time period, the SDK MUST reject the request and return a `500 Internal Server Error` response.
+5. The SDK then MUST calculate an HMAC with SHA256 by specifying the key as the `s` value of the `X-Inngest-Signature` header with the `signkey-*-` prefix removed and hashing the body of the request followed by the timestamp `t` with hex encoding.
+
+Note that if the raw bytes of the request body are inaccessible, the body should first be parsed using the JSON Canonicalization Scheme (JCS) as specified in [RFC 8785](https://datatracker.ietf.org/doc/html/rfc8785).
+
+If the generated hash does not exactly match the `s` value in the `X-Inngest-Signature` header, the SDK MUST reject the requets with a `500 Internal Server Error`.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:235]:
```suggestion
If the generated hash does not exactly match the `s` value in the `X-Inngest-Signature` header, the SDK MUST reject the requests with a `500 Internal Server Error`.
```

File: docs/SDK_SPEC.md [hunk 3]
```diff
@@ -0,0 +1,1148 @@
...
+
+# header
+Authorization: Bearer signkey-prod-3c8335d113497a3a0b3e6bc18c12bf59e3db1c964c9f66765f374f5f7b473ac7
+```
+
+## 4.2. Kinds of Inngest Server
+
+Inngest provides both a hosted Cloud (production) service and a Dev Server which allows local development of Inngest Functions. We refer to both as an Inngest Server.
+
+### 4.2.1. Targetting an Inngest Server
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:262]:
```suggestion
### 4.2.1. Targeting an Inngest Server
```

File: docs/SDK_SPEC.md [hunk 4]
```diff
@@ -0,0 +1,1148 @@
...
+If the `INNGEST_DEV` value is a vcalid [RFC 6454](https://datatracker.ietf.org/doc/html/rfc6454) origin, for example `http://example.com`, the resulting endpoints will then be used:
+
+- API Origin: `http://example.com`
+- Event ingestion endpoint: `http://example.com/e/:event_key`
+
+### 4.2.2. Request verification
+
+When receiving requests expected to be from an Inngest Server, an SDK MUST verify their integrity [4.1.3]. An Inngest Dev Server, however, MUST NOT require verification.
+
+If the recommendations when targetting an Inngest Server [4.2.1] are followed, then presence of the `INNGEST_DEV` environment variable (or equivalent runtime configuration option) MUST turn off signature verification.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:292]:
```suggestion
If the recommendations when targeting an Inngest Server [4.2.1] are followed, then presence of the `INNGEST_DEV` environment variable (or equivalent runtime configuration option) MUST turn off signature verification.
```

File: docs/SDK_SPEC.md [hunk 5]
```diff
@@ -0,0 +1,1148 @@
...
+
+### 4.2.3. Proxies and routing
+
+The URL through which an SDK communicates with an Inngest Server may be indirect. In these cases, an SDK SHOULD provide the Developer with a method of setting the origin of API requests and Event ingestion.
+
+We RECOMMEND that a Developer an use the `INNGEST_API_ORIGIN` and `INNGEST_EVENT_API_ORIGIN` environment variables [3.2] for this.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:298]:
```suggestion
We RECOMMEND that a Developer can use the `INNGEST_API_ORIGIN` and `INNGEST_EVENT_API_ORIGIN` environment variables [3.2] for this.
```

Guessing

File: docs/SDK_SPEC.md [hunk 6]
```diff
@@ -0,0 +1,1148 @@
...
+    Inngest-->>-SDK: 200
+    SDK-->>-?: 200
+end
+```
+
+The numbered statements in the diagram above will be detailed in the following sections.
+
+### 4.3.1. Receiving a Sync Request
+
+When an SDK receives a `PUT` request, it is treated as the first step of a Sync. This request can come from automatically from Inngest, from being triggered when using an Inngest UI, or from any `PUT` request, allowing a Developer to manually trigger a deploy by contacting their own service.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:329]:
```suggestion
When an SDK receives a `PUT` request, it is treated as the first step of a Sync. This request can come automatically from Inngest, from being triggered when using an Inngest UI, or from any `PUT` request, allowing a Developer to manually trigger a deploy by contacting their own service.
```

File: docs/SDK_SPEC.md [hunk 7]
```diff
@@ -0,0 +1,1148 @@
...
+
+    /**
+     * Allow the specification of an idempotency key using event data. If
+     * specified, this overrides the `rateLimit` object.
+     */
+    idempotency?: string;
+
+    /**
+     * batchEvents specifies the batch configuration on when this function
+     * should be invoked when one of the requirements are fulfilled.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:482]:
```suggestion
     * should be invoked when one of the requirements is fulfilled.
```

File: docs/SDK_SPEC.md [hunk 8]
```diff
@@ -0,0 +1,1148 @@
...
+       * - async, referencing the new cancel event.
+       */
+      if?: string;
+
+      /**
+       * An optional timeout that the cancel is valid for.  If this isn't
+       * specified, cancellation triggers are valid for up to a year or until the
+       * function ends.
+       *
+       * The time to wait can be specifed using a Time String or an ISO 8601
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:520]:
```suggestion
       * The time to wait can be specified using a Time String or an ISO 8601
```

File: docs/SDK_SPEC.md [hunk 9]
```diff
@@ -0,0 +1,1148 @@
...
+       */
+      period: TimeStr;
+    };
+
+    /**
+     * Debounce delays functions for the `period` specified. If an event is sent,
+     * the function will not run until at least `period` has elapsed.
+     *
+     * If any new events are received that match the same debounce `key`, the
+     * function is reshceduled for another `period` delay, and the triggering
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:553]:
```suggestion
     * function is rescheduled for another `period` delay, and the triggering
```

File: docs/SDK_SPEC.md [hunk 10]
```diff
@@ -0,0 +1,1148 @@
...
+
+The SDK MUST adhere to the global header requirements when sending requests [4.1.4].
+
+If an `X-Inngest-Server-Kind` header was set on the sync request received by the SDK, the SDK SHOULD forward the header’s value with a new name of `X-Inngest-Expected-Server-Kind`. This header is used to combat situations where services that a Developer understands to target Inngest Cloud can not be synced with an Inngest Dev Server and vice versa.
+
+### 4.3.3. Handling failure
+
+If Inngest returns a non-`200 OK` status code, the Sync has failed.
+
+A `400 Bad Request` response indicates that either the App and Function configurations passed were invalid, or the Inngest service reaached was not the one that started the deployment, for example if the Inngest Dev Server starts the deployment, but the SDK reaches out to Inngest Cloud.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:644]:
```suggestion
A `400 Bad Request` response indicates that either the App and Function configurations passed were invalid, or the Inngest service reached was not the one that started the deployment, for example if the Inngest Dev Server starts the deployment, but the SDK reaches out to Inngest Cloud.
```

File: docs/SDK_SPEC.md [hunk 11]
```diff
@@ -0,0 +1,1148 @@
...
+	message: "Successfully synced.",
+	modified: true;
+}
+```
+
+## 4.4. Call Requests
+
+A Call Request is a single request from Inngest to an SDK in order to discover work that must be queued or to run blocks of user code defined in Functions or Steps.
+
+Over the course of an executing a Function that makes no use of Steps, the happy path sees Inngest execute a Function only once, where the Function’s contents are run during that Execution and the result returned to Inngest. If this Function fails by user code throwing an error or some external issue such as network issues or out-of-memory errors, Inngest will retry the Function, by executing it again for each retry.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:691]:
```suggestion
Over the course of executing a Function that makes no use of Steps, the happy path sees Inngest execute a Function only once, where the Function’s contents are run during that Execution and the result returned to Inngest. If this Function fails by user code throwing an error or some external issue such as network issues or out-of-memory errors, Inngest will retry the Function, by executing it again for each retry.
```

File: docs/SDK_SPEC.md [hunk 12]
```diff
@@ -0,0 +1,1148 @@
...
+If the request was successful, the returned Step data should be used to memoize discovered Steps. If the request failed or returned a non-`200 OK` status code, the SDK MUST respond to the Call Request with a `500 Internal Server Error`.
+
+### 4.4.3. Executing the Function
+
+When an SDK has now validated the request is from Inngest and has retrieved the entire payload, it is ready to run the code represented by the Function.
+
+An SDK SHOULD provide the Developer’s code with select fields from the Call Request’s payload, ensuring that a Developer can utilize these fields to make appropriate decisions in code and change functionality based on the situation. We RECOMMEND that these values are provided, with casing changed to the most idiomatic casing used in a particular SDK:
+
+- `event` - The entire JSON object passed to the Execution.
+- `events` - The entire array os JSON objects passed to the Execution.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:817]:
```suggestion
- `events` - The entire array of JSON objects passed to the Execution.
```

File: docs/SDK_SPEC.md [hunk 13]
```diff
@@ -0,0 +1,1148 @@
...
+- `runId` - Accessed from `ctx.run_id`.
+- `attempt` - Accessed from `ctx.attempt`.
+- `step` - Step tooling as detailed in Steps [5].
+
+An SDK MUST then call the Function and return either the returned value as JSON or an error to Inngest. When responding, an SDK MUST adhere to the requirements of responding to a request from an Inngest Server [4.1.2].
+
+A successful response to Inngest MUST always return `200 OK` with a a JSON payload which is the return value of the function.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:824]:
```suggestion
A successful response to Inngest MUST always return `200 OK` with a JSON payload which is the return value of the function.
```

File: docs/SDK_SPEC.md [hunk 14]
```diff
@@ -0,0 +1,1148 @@
...
+- `id` is the hashed ID of the Step
+- `op` is the string “opcode” (an abbreviation of “operation code”) that specifies the type of Step being reported. The functionality and data expectations of each Step type is different. See Available Step types [5.3].
+- `displayName` is an optional string that can be given to provide a human-readable name to the Step in UIs and logging. If this is not specified by the user, it is RECOMMENDED that an SDK set this to the ID for the step that was provided instead.
+- `opts` is an object used across many Step types to fine tune their functionality. For information on the expectations of this object for each Step type, see Available Step types [5.3].
+
+### 5.1.2. IDs and hashing
+
+The `id` of each Step MUST be a hex-encoded SHA-1 hash such that the Step can be identified reliably during multiple Call Requests. Commonly, this is done by hashing a human-readable identifier provided by the Developer.
+
+For example, the human-readable identififer `my-step-id` would be hashed as `e7d8a2f140845095749d60246ff1110c9d01d76a`.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:923]:
```suggestion
For example, the human-readable identifier `my-step-id` would be hashed as `e7d8a2f140845095749d60246ff1110c9d01d76a`.
```

File: docs/SDK_SPEC.md [hunk 15]
```diff
@@ -0,0 +1,1148 @@
...
+
+```tsx
+{
+	name: string;
+	message: string;
+	stack?: string;
+}
+```
+
+`null` can be also be found depending on the Step type [5.3], where the particular Step type being used indicates its meaning.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:957]:
```suggestion
`null` can also be found depending on the Step type [5.3], where the particular Step type being used indicates its meaning.
```

File: docs/SDK_SPEC.md [hunk 16]
```diff
@@ -0,0 +1,1148 @@
...
+
+If an `error` is returned/thrown/raised and the Developer has not either swallowed the error or returned/thrown/raised a new error, the SDK MUST mark the request as non-retriable [4.4.3].
+
+If a Step’s hash is present in the `steps` object passed in the Call Request, it MUST NOT be reported when responding to that Call Request; only new Steps without a memoized `data` or `error` can be reported.
+
+If a Step is found to have a memoized `error`, an SDK SHOULD create the equivalent of a `StepError` value, passed to the Developer such that they can use it to access the `name`, `message`, and optional `stack` property.
+
+The `StepError` value SHOULD aim to be used as a language-native error, e.g. JavaScript should extend `Error`, Python should extend `Exception`, and Go should implement the `error` interface. This ensures that the Developer can utilize language features such as `error.As()` r `except StepError as e` to appropriately catch and type the error.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:969]:
```suggestion
The `StepError` value SHOULD aim to be used as a language-native error, e.g. JavaScript should extend `Error`, Python should extend `Exception`, and Go should implement the `error` interface. This ensures that the Developer can utilize language features such as `error.As()` or `except StepError as e` to appropriately catch and type the error.
```

File: docs/SDK_SPEC.md [hunk 17]
```diff
@@ -0,0 +1,1148 @@
...
+	id: string;
+	op: "StepNotFound";
+}
+```
+
+The memoized result of the Step will be either a `{ data }` or an `{ error }` object, depending on what was passed by the SDK.
+
+### 5.3.2. Sleep
+
+A Sleep Step informs the Inngest Server that the Run wishes to be called again after the given amount of time, specified either as an ISO 8601 date as specified [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339), or a “time string.” The latter “time string” is a sequence of a decimal numbers, each with an optional fraction and unit suffix, such as `"300ms"`, or `"2h45m"`. Valid time units are `"ns"`, `"us"` (or `"µs"`), `"ms"`, `"s"`, `"m"`, `"h"`, `"d"`, and `"w"`.
```
Comment (sylwiavargas) [inline docs/SDK_SPEC.md:1071]:
```suggestion
A Sleep Step informs the Inngest Server that the Run wishes to be called again after the given amount of time, specified either as an ISO 8601 date as specified [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339), or a “time string.” The latter “time string” is a sequence of decimal numbers, each with an optional fraction and unit suffix, such as `"300ms"`, or `"2h45m"`. Valid time units are `"ns"`, `"us"` (or `"µs"`), `"ms"`, `"s"`, `"m"`, `"h"`, `"d"`, and `"w"`.
```

**Review by sylwiavargas — APPROVED** [review]
Such a great document. Thank you.
Can I add it to my GoodReads as a book of the week?

**Review by djfarrelly — APPROVED** [review]

**Pull request merged**